In [1]:
import os


os.chdir("../")


os.getcwd()


'd:\\DA_workspace\\heart_disease_predictor\\src'

In [2]:
import pandas as pd
import numpy as np
from classifier.Mylib import myfuncs
from sklearn.compose import ColumnTransformer
from sklearn.impute import SimpleImputer
from sklearn.base import BaseEstimator, TransformerMixin
from sklearn.model_selection import train_test_split


Doc du lieu


In [3]:
df = myfuncs.load_python_object("../artifacts/data_ingestion/train_data.pkl")

df.head()


State     Sex GeneralHealth  PhysicalHealthDays  MentalHealthDays  \
0   New York    Male          Poor                 0.0              20.0   
1  Minnesota    Male          Good                 0.0               0.0   
2       Ohio    Male     Excellent                 0.0               1.0   
3    Montana  Female          Good                 1.0               0.0   
4  Minnesota  Female          Fair                 2.0               7.0   

                                     LastCheckupTime PhysicalActivities  \
0  Within past year (anytime less than 12 months ...                 No   
1  Within past 2 years (1 year but less than 2 ye...                Yes   
2  Within past 2 years (1 year but less than 2 ye...                Yes   
3  Within past year (anytime less than 12 months ...                Yes   
4  Within past year (anytime less than 12 months ...                Yes   

   SleepHours  RemovedTeeth HadHeartAttack  ... HeightInMeters  \
0         5.0        1 to 5            Yes  ...           1.83   
1         6.0  None of them             No  ...           1.80   
2         7.0  None of them             No  ...           1.68   
3         8.0           All             No  ...           1.65   
4         6.0        1 to 5            Yes  ...           1.63   

  WeightInKilograms    BMI AlcoholDrinkers HIVTesting FluVaxLast12  \
0            113.40  33.91             Yes        Yes           No   
1             99.79  30.68             Yes        Yes           No   
2             60.78  21.63             Yes         No          Yes   
3             98.88  36.28             Yes        Yes          Yes   
4             84.82  32.10              No         No           No   

  PneumoVaxEver                                  TetanusLast10Tdap  \
0            No  Yes, received tetanus shot but not sure what type   
1            No  Yes, received tetanus shot but not sure what type   
2           Yes  Yes, received tetanus shot but not sure what type   
3           Yes  Yes, received tetanus shot but not sure what type   
4           Yes           Yes, received tetanus shot, but not Tdap   

  HighRiskLastYear                                           CovidPos  
0               No  Tested positive using home test without a heal...  
1               No  Tested positive using home test without a heal...  
2               No                                                 No  
3               No                                                 No  
4               No                                                Yes  

[5 rows x 40 columns]

Kich thuoc


In [4]:
df.shape


(397860, 40)

Ti le missing cac cot


In [7]:
null_percent =  df.isnull().mean() * 100
null_percent = null_percent.sort_values(ascending=False)
null_percent


TetanusLast10Tdap            18.452471
PneumoVaxEver                17.230936
HIVTesting                   14.773036
ChestScan                    12.505907
CovidPos                     11.323078
HighRiskLastYear             11.297190
BMI                          10.882220
FluVaxLast12                 10.501935
AlcoholDrinkers              10.371990
WeightInKilograms             9.376414
ECigaretteUsage               7.940230
SmokerStatus                  7.887699
HeightInMeters                6.370080
DifficultyErrands             5.692706
DifficultyConcentrating       5.369477
DifficultyWalking             5.324235
DifficultyDressingBathing     5.303876
BlindOrVisionDifficulty       4.787362
DeafOrHardOfHearing           4.579249
RaceEthnicityCategory         3.100588
RemovedTeeth                  2.499623
PhysicalHealthDays            2.392299
AgeCategory                   2.001709
MentalHealthDays              1.984869
LastCheckupTime               1.818479
SleepHours               

Ti le missing cua cot ma lon nhat la nho hon 19% nen Khong xoa cot nao het !


Kich thuoc con lai


In [8]:
df.shape


(397860, 40)

In [23]:
df['TetanusLast10Tdap'].unique()


array(['Yes, received tetanus shot but not sure what type',
       'Yes, received tetanus shot, but not Tdap', nan,
       'No, did not receive any tetanus shot in the past 10 years',
       'Yes, received Tdap'], dtype=object)

Y nghia cac cot

| Cot                       | Y nghia                                                               | Don vi | Phan loai |
| ------------------------- | --------------------------------------------------------------------- | ------ | --------- |
| State                     | bang                                                                  | none   | Nominal   |
| Sex                       | gioi tinh                                                             | none   | Nominal   |
| GeneralHealth             | suc khoe chung                                                        | none   | ordinal   |
| PhysicalHealthDays        | so ngay gap van de the chat                                           | none   | numeric   |
| MentalHealthDays          | so ngay gap van de tinh than                                          | none   | numeric   |
| LastCheckupTime           | lan cuoi di kham suc khoe                                             | none   | ordinal   |
| PhysicalActivities        | co tham gia hoat dong the chat khong                                  | none   | binary    |
| SleepHours                | gio ngu trong 1 ngay                                                  | none   | numeric   |
| RemovedTeeth              | so rang da bi loai bo                                                 | none   | nominal   |
| **HadHeartAttack**        | co bi heart attack khong                                              | none   | binary    |
| HadAngina                 | co bi đau thắt ngực                                                   | none   | binary    |
| HadStroke                 | co bi đột quỵ                                                         | none   | binary    |
| HadAsthma                 | co bi hen suyen khong                                                 | none   | binary    |
| HadSkinCancer             | co bi ung thu da khong                                                | none   | binary    |
| HadCOPD                   | co bi Bệnh phổi tắc nghẽn mãn tính                                    | none   | binary    |
| HadDepressiveDisorder     | co bi roi loan tam than khong                                         | none   | binary    |
| HadKidneyDisease          | co bi benh than khong                                                 | none   | binary    |
| HadArthritis              | co bi viêm khớp khong                                                 | none   | binary    |
| HadDiabetes               | co bi tieu duong khong                                                | none   | binary    |
| DeafOrHardOfHearing       | co bi chung kho nghe khong                                            | none   | binary    |
| BlindOrVisionDifficulty   | co bi chung kho nhin khong                                            | none   | binary    |
| DifficultyConcentrating   | co bi chung kho tap trung khong                                       | none   | binary    |
| DifficultyWalking         | co bi chung kho di lai khong                                          | none   | binary    |
| DifficultyDressingBathing | co bi gap kho khan trong viec thay do hay tam rua khong               | none   | binary    |
| DifficultyErrands         | co bi gap kho khan trong viec thuc hien cac cong viec hang ngay khong | none   | binary    |
| SmokerStatus              | muc do hut thuoc truyen thong                                         | none   | ordinal   |
| ECigaretteUsage           | muc do hut thuoc la dien tu                                           | none   | ordinal   |
| _ChestScan_               |                                                                       | none   | remove    |
| _RaceEthnicityCategory_   | chung toc                                                             | none   | remove    |
| AgeCategory               | pham vi tuoi                                                          | none   | nominal   |
| HeightInMeters            | chieu cao                                                             | m      | numeric   |
| WeightInKilograms         | can nang                                                              | kg     | numeric   |
| _BMI_                     | chi so BMI                                                            | none   | remove    |
| AlcoholDrinkers           | co uong ruou khong                                                    | none   | binary    |
| _HIVTesting_              |                                                                       | none   | remove    |
| FluVaxLast12              | đã tiêm vắc xin phòng cúm trong 12 tháng qua hay không.               | none   | binary    |
| PneumoVaxEver             | đã từng tiêm vắc xin phòng phổi (Pneumococcal vaccine) hay chưa       | none   | binary    |
| TetanusLast10Tdap         | muc do tiêm vắc xin phòng uốn ván trong 10 nam qua                    | none   | nominal   |
| _HighRiskLastYear_        |                                                                       | none   | remove    |
| _CovidPos_                |                                                                       | none   | remove    |

PhysicalHealthDays: vd: Nếu một người báo cáo rằng trong 30 ngày qua, họ đã có 5 ngày bị ảnh hưởng bởi cơn đau lưng và không thể hoạt động bình thường, thì giá trị trong cột PhysicalHealthDays la 5.

MentalHealthDays: vd: Nếu một người báo cáo rằng trong 30 ngày qua, họ đã có 5 ngày bị ảnh hưởng bởi tinh than và khong co cam xuc binh thuong, thì giá trị trong cột MentalHealthDays la 5.

Cong thuc tinh $BMI = canNang / chieuCao^2$ -> cac cot HeightInMeters, WeightInKilograms, BMI can loai bo 1 cot di


Xoa cac cot khong co y nghia


In [24]:
df = df.drop(columns = ['ChestScan', 'RaceEthnicityCategory', 'BMI', 'HIVTesting', 'HighRiskLastYear', 'CovidPos'])


Doi ten cot


In [25]:
df.columns


Index(['State', 'Sex', 'GeneralHealth', 'PhysicalHealthDays',
       'MentalHealthDays', 'LastCheckupTime', 'PhysicalActivities',
       'SleepHours', 'RemovedTeeth', 'HadHeartAttack', 'HadAngina',
       'HadStroke', 'HadAsthma', 'HadSkinCancer', 'HadCOPD',
       'HadDepressiveDisorder', 'HadKidneyDisease', 'HadArthritis',
       'HadDiabetes', 'DeafOrHardOfHearing', 'BlindOrVisionDifficulty',
       'DifficultyConcentrating', 'DifficultyWalking',
       'DifficultyDressingBathing', 'DifficultyErrands', 'SmokerStatus',
       'ECigaretteUsage', 'AgeCategory', 'HeightInMeters', 'WeightInKilograms',
       'AlcoholDrinkers', 'FluVaxLast12', 'PneumoVaxEver',
       'TetanusLast10Tdap'],
      dtype='object')

In [ ]:
rename_dict = {
    "State": "State_nom",
    "Sex": "Sex_nom",
    "GeneralHealth": "GeneralHealth_ord",
    "PhysicalHealthDays": "PhysicalHealthDays_num",
    "MentalHealthDays": "MentalHealthDays_num",
    "LastCheckupTime": "LastCheckupTime_ord",
    "LastCheckupTime": "LastCheckupTime_ord",
}

nnnnnn

df = df.rename(columns=rename_dict)

df.head()


Country  Age  Gender  Education_Level   BMI Physical_Activity_Level  \
0        France   52  Female                6  21.4                    High   
1         Spain   62  Female                4  28.7                     Low   
2     Australia   52    Male               19  21.3                  Medium   
3        Norway   74    Male               15  32.4                     Low   
4  Saudi Arabia   94  Female               17  33.9                     Low   

  Smoking_Status_level Alcohol_Consumption_Level do_get_Diabetes  \
0              Current              Occasionally              No   
1              Current              Occasionally              No   
2              Current              Occasionally              No   
3              Current                     Never              No   
4               Former                 Regularly              No   

  do_get_Hypertension  ... Diet_level Air_Pollution_Exposure_level  \
0                  No  ...    Healthy                         High   
1                 Yes  ...    Average                         High   
2                  No  ...  Unhealthy                       Medium   
3                  No  ...    Average                          Low   
4                 Yes  ...  Unhealthy                       Medium   

   Employment_Status Marital_Status do_get_Genetic_Risk_Factor  \
0           Employed        Married                        Yes   
1           Employed        Widowed                         No   
2            Retired         Single                         No   
3         Unemployed        Married                         No   
4            Retired        Widowed                         No   

  Social_Engagement_Level Income_Level Stress_Levels  \
0                    High         High          High   
1                     Low       Medium        Medium   
2                    High       Medium           Low   
3                     Low         High          High   
4                  Medium         High          High   

  living_place_urban_or_rural do_get_Alzheimer  
0                       Rural              Yes  
1                       Rural               No  
2                       Urban               No  
3                       Rural               No  
4                       Rural              Yes  

[5 rows x 25 columns]

chuyen lai ten cot co hau to nhu sau

- cot so: num
- cot binary: bin
- cot nominal: nom
- cot ordinal: ord
-


In [ ]:
rename_dict = {
    "Age": "Age_num",
    "Education_Level": "Education_Level_num",
    "BMI": "BMI_num",
    "Cognitive_Test_Score": "Cognitive_Test_Score_num",
    "do_get_Diabetes": "do_get_Diabetes_bin",
    "do_get_Hypertension": "do_get_Hypertension_bin",
    "do_Family_History": "do_Family_History_bin",
    "do_get_Genetic_Risk_Factor": "do_get_Genetic_Risk_Factor_bin",
    "Gender": "Gender_nom",
    "Country": "Country_nom",
    "Employment_Status": "Employment_Status_nom",
    "Marital_Status": "Marital_Status_nom",
    "living_place_urban_or_rural": "living_place_urban_or_rural_nom",
    "Physical_Activity_Level": "Physical_Activity_Level_ord",
    "Smoking_Status_level": "Smoking_Status_level_ord",
    "Alcohol_Consumption_Level": "Alcohol_Consumption_Level_ord",
    "Cholesterol_Level": "Cholesterol_Level_ord",
    "Depression_Level": "Depression_Level_ord",
    "Sleep_Quality_Level": "Sleep_Quality_Level_ord",
    "Diet_level": "Diet_level_ord",
    "Air_Pollution_Exposure_level": "Air_Pollution_Exposure_level_ord",
    "Social_Engagement_Level": "Social_Engagement_Level_ord",
    "Income_Level": "Income_Level_ord",
    "Stress_Levels": "Stress_Levels_ord",
    "Stress_Levels": "Stress_Levels_ord",
}

df = df.rename(columns=rename_dict)

df.columns


Index(['Country_nom', 'Age_num', 'Gender_nom', 'Education_Level_num',
       'BMI_num', 'Physical_Activity_Level_ord', 'Smoking_Status_level_ord',
       'Alcohol_Consumption_Level_ord', 'do_get_Diabetes_bin',
       'do_get_Hypertension_bin', 'Cholesterol_Level_ord',
       'do_Family_History_bin', 'Cognitive_Test_Score_num',
       'Depression_Level_ord', 'Sleep_Quality_Level_ord', 'Diet_level_ord',
       'Air_Pollution_Exposure_level_ord', 'Employment_Status_nom',
       'Marital_Status_nom', 'do_get_Genetic_Risk_Factor_bin',
       'Social_Engagement_Level_ord', 'Income_Level_ord', 'Stress_Levels_ord',
       'living_place_urban_or_rural_nom', 'do_get_Alzheimer'],
      dtype='object')

In [ ]:
df.columns


Index(['Country_nom', 'Age_num', 'Gender_nom', 'Education_Level_num',
       'BMI_num', 'Physical_Activity_Level_ord', 'Smoking_Status_level_ord',
       'Alcohol_Consumption_Level_ord', 'do_get_Diabetes_bin',
       'do_get_Hypertension_bin', 'Cholesterol_Level_ord',
       'do_Family_History_bin', 'Cognitive_Test_Score_num',
       'Depression_Level_ord', 'Sleep_Quality_Level_ord', 'Diet_level_ord',
       'Air_Pollution_Exposure_level_ord', 'Employment_Status_nom',
       'Marital_Status_nom', 'do_get_Genetic_Risk_Factor_bin',
       'Social_Engagement_Level_ord', 'Income_Level_ord', 'Stress_Levels_ord',
       'living_place_urban_or_rural_nom', 'do_get_Alzheimer'],
      dtype='object')

Y nghia cac cot

| Cot                          | Y nghia                                | Don vi | Phan loai |
| ---------------------------- | -------------------------------------- | ------ | --------- |
| Country                      | Dat nuoc dang song                     | none   | Nominal   |
| Age                          | Tuoi                                   | none   | Numeric   |
| Gender                       | Gioi tinh                              | none   | Numeric   |
| Education_Level              | Muc do hoc van                         | none   | Numeric   |
| BMI                          | Chi so BMI (18.5 - 35)                 | none   | Numeric   |
| Physical_Activity_Level      | Muc do hoat dong the chat              | none   | Ordinal   |
| Smoking_Status_level         | Trang thai hut thuoc                   | none   | Ordinal   |
| Alcohol_Consumption_Level    | Tan suat tieu thu ruou                 | none   | Ordinal   |
| do_get_Diabetes              | Co bi tieu duong khong                 | none   | binary    |
| do_get_Hypertension          | Co bi huyet ap cao khong               | none   | binary    |
| Cholesterol_Level            | Muc do Cholesterol                     | none   | ordinal   |
| do_Family_History            | Co nguoi nha bi Alzheimer’s khong      | none   | binary    |
| Cognitive_Test_Score         | Diem kiem tra tu duy                   | none   | Numeric   |
| Depression_Level             | Muc do tram cam                        | none   | ordinal   |
| Sleep_Quality_Level          | Chat luong giac ngu                    | none   | ordinal   |
| Diet_level                   | Chat luong bua an                      | none   | ordinal   |
| Air_Pollution_Exposure_level | Muc do tiep xuc voi o nhiem khong khi  | none   | ordinal   |
| Employment_Status            | Trang thai viec lam hien tai           | none   | nominal   |
| Marital_Status               | Trang thai hon nhan                    | none   | nominal   |
| do_get_Genetic_Risk_Factor   | Co dinh yeu to di truyen khong         | none   | binary    |
| Social_Engagement_Level      | muc do tham gia vao cong tac xa hoi    | none   | ordinal   |
| Income_Level                 | muc do thu nhap                        | none   | ordinal   |
| Stress_Levels                | muc do stress                          | none   | ordinal   |
| living_place_urban_or_rural  | noi song                               | none   | nominal   |
| do_get_Alzheimer             | bien muc tieu, co bi Alzheimer’s khong | none   | binary    |

Marital_Status = "Widowed":có nghĩa là góa, tức là một người đã mất vợ hoặc chồng.

**do_get_Alzheimer**

- Yes: Co bi Alzheimer
- No: Khong bi Alzheimer


Chuyen cac cot numeric, binary, nominal, ordinal, target ve dung vi tri


In [ ]:
df.columns


Index(['Country_nom', 'Age_num', 'Gender_nom', 'Education_Level_num',
       'BMI_num', 'Physical_Activity_Level_ord', 'Smoking_Status_level_ord',
       'Alcohol_Consumption_Level_ord', 'do_get_Diabetes_bin',
       'do_get_Hypertension_bin', 'Cholesterol_Level_ord',
       'do_Family_History_bin', 'Cognitive_Test_Score_num',
       'Depression_Level_ord', 'Sleep_Quality_Level_ord', 'Diet_level_ord',
       'Air_Pollution_Exposure_level_ord', 'Employment_Status_nom',
       'Marital_Status_nom', 'do_get_Genetic_Risk_Factor_bin',
       'Social_Engagement_Level_ord', 'Income_Level_ord', 'Stress_Levels_ord',
       'living_place_urban_or_rural_nom', 'do_get_Alzheimer'],
      dtype='object')

In [ ]:
cols = pd.Series(df.columns)
numeric_cols = cols[cols.str.endswith("num")].tolist()
binary_cols = cols[cols.str.endswith("bin")].tolist()
nominal_cols = cols[cols.str.endswith("nom")].tolist()
ordinal_cols = cols[cols.str.endswith("ord")].tolist()

target_col = "do_get_Alzheimer"



df = df[numeric_cols + binary_cols + nominal_cols + ordinal_cols + [target_col]]



df.head()


Age_num  Education_Level_num  BMI_num  Cognitive_Test_Score_num  \
0       52                    6     21.4                        67   
1       62                    4     28.7                        73   
2       52                   19     21.3                        91   
3       74                   15     32.4                        70   
4       94                   17     33.9                        76   

  do_get_Diabetes_bin do_get_Hypertension_bin do_Family_History_bin  \
0                  No                      No                    No   
1                  No                     Yes                   Yes   
2                  No                      No                    No   
3                  No                      No                    No   
4                  No                     Yes                   Yes   

  do_get_Genetic_Risk_Factor_bin   Country_nom Gender_nom  ...  \
0                            Yes        France     Female  ...   
1                             No         Spain     Female  ...   
2                             No     Australia       Male  ...   
3                             No        Norway       Male  ...   
4                             No  Saudi Arabia     Female  ...   

  Alcohol_Consumption_Level_ord Cholesterol_Level_ord Depression_Level_ord  \
0                  Occasionally                Normal               Medium   
1                  Occasionally                Normal                  Low   
2                  Occasionally                  High                  Low   
3                         Never                Normal               Medium   
4                     Regularly                  High               Medium   

  Sleep_Quality_Level_ord Diet_level_ord Air_Pollution_Exposure_level_ord  \
0                    Poor        Healthy                             High   
1                    Good        Average                             High   
2                    Good      Unhealthy                           Medium   
3                 Average        Average                              Low   
4                    Good      Unhealthy                           Medium   

  Social_Engagement_Level_ord Income_Level_ord Stress_Levels_ord  \
0                        High             High              High   
1                         Low           Medium            Medium   
2                        High           Medium               Low   
3                         Low             High              High   
4                      Medium             High              High   

  do_get_Alzheimer  
0              Yes  
1               No  
2               No  
3               No  
4              Yes  

[5 rows x 25 columns]

Kiem tra kieu du lieu cac cot


In [ ]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66854 entries, 0 to 66853
Data columns (total 25 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Age_num                           66854 non-null  int64  
 1   Education_Level_num               66854 non-null  int64  
 2   BMI_num                           66854 non-null  float64
 3   Cognitive_Test_Score_num          66854 non-null  int64  
 4   do_get_Diabetes_bin               66854 non-null  object 
 5   do_get_Hypertension_bin           66854 non-null  object 
 6   do_Family_History_bin             66854 non-null  object 
 7   do_get_Genetic_Risk_Factor_bin    66854 non-null  object 
 8   Country_nom                       66854 non-null  object 
 9   Gender_nom                        66854 non-null  object 
 10  Employment_Status_nom             66854 non-null  object 
 11  Marital_Status_nom                66854 non-null  object 
 12  livi

DUNG

- Age
- Education_Level
- BMI
- Cognitive_Test_Score

---

SAI
|Cot|Hien tai | Phai la|
|---|---|---|
|do_get_Diabetes | object | int |
|do_get_Hypertension | object | int |
|do_Family_History | object | int |
|do_get_Genetic_Risk_Factor | object | int |
|Gender | object | category |
|Country | object | category |
|Employment_Status | object | category |
|Marital_Status | object | category |
|living_place_urban_or_rural | object | category |
|Physical_Activity_Level | object | category |
|Smoking_Status_level | object | category |
|Alcohol_Consumption_Level | object | category |
|Cholesterol_Level | object | category |
|Depression_Level | object | category |
|Sleep_Quality_Level | object | category |
|Diet_level | object | category |
|Air_Pollution_Exposure_level | object | category |
|Social_Engagement_Level | object | category |
|Income_Level | object | category |
|Stress_Levels | object | category |
|do_get_Alzheimer | object | category |


Kiem tra cac cot co kdl = kdlmm + NA


In [ ]:
set(map(type, df["do_get_Diabetes_bin"]))


{str}

In [ ]:
set(map(type, df["do_get_Hypertension_bin"]))


{str}

In [ ]:
set(map(type, df["do_Family_History_bin"]))


{str}

In [ ]:
set(map(type, df["do_get_Genetic_Risk_Factor_bin"]))


{str}

In [ ]:
set(map(type, df["Gender_nom"]))


{str}

In [ ]:
set(map(type, df["Country_nom"]))


{str}

In [ ]:
set(map(type, df["Employment_Status_nom"]))


{str}

In [ ]:
set(map(type, df["Marital_Status_nom"]))


{str}

In [ ]:
set(map(type, df["living_place_urban_or_rural_nom"]))


{str}

In [ ]:
set(map(type, df["Physical_Activity_Level_ord"]))


{str}

In [ ]:
set(map(type, df["Smoking_Status_level_ord"]))


{str}

In [ ]:
set(map(type, df["Alcohol_Consumption_Level_ord"]))


{str}

In [ ]:
set(map(type, df["Cholesterol_Level_ord"]))


{str}

In [ ]:
set(map(type, df["Depression_Level_ord"]))


{str}

In [ ]:
set(map(type, df["Sleep_Quality_Level_ord"]))


{str}

In [ ]:
set(map(type, df["Diet_level_ord"]))


{str}

In [ ]:
set(map(type, df["Air_Pollution_Exposure_level_ord"]))


{str}

In [ ]:
set(map(type, df["Social_Engagement_Level_ord"]))


{str}

In [ ]:
set(map(type, df["Income_Level_ord"]))


{str}

In [ ]:
set(map(type, df["Stress_Levels_ord"]))


{str}

In [ ]:
set(map(type, df["do_get_Alzheimer"]))


{str}

Kiem tra noi dung cac cot

Cot numeric


In [ ]:
df[numeric_cols].describe().loc[["min", "max", "mean"]]


Age_num  Education_Level_num    BMI_num  Cognitive_Test_Score_num
min   50.000000             0.000000  18.500000                 30.000000
max   94.000000            19.000000  35.000000                 99.000000
mean  71.930969             9.489918  26.781271                 64.655503

Cot binary


In [ ]:
for col in binary_cols:
    print(f"{col} -> {df[col].unique().tolist()}")


do_get_Diabetes_bin -> ['No', 'Yes']
do_get_Hypertension_bin -> ['No', 'Yes']
do_Family_History_bin -> ['No', 'Yes']
do_get_Genetic_Risk_Factor_bin -> ['Yes', 'No']


Cot nominal


In [ ]:
for col in nominal_cols:
    print(f"{col} -> {df[col].unique().tolist()}")


Country_nom -> ['France', 'Spain', 'Australia', 'Norway', 'Saudi Arabia', 'South Korea', 'Japan', 'Germany', 'Mexico', 'Argentina', 'China', 'USA', 'Brazil', 'Russia', 'Sweden', 'South Africa', 'India', 'Italy', 'UK', 'Canada']
Gender_nom -> ['Female', 'Male']
Employment_Status_nom -> ['Employed', 'Retired', 'Unemployed']
Marital_Status_nom -> ['Married', 'Widowed', 'Single']
living_place_urban_or_rural_nom -> ['Rural', 'Urban']


Cot ordinal


In [ ]:
for col in ordinal_cols:
    print(f"{col} -> {df[col].unique().tolist()}")


Physical_Activity_Level_ord -> ['High', 'Low', 'Medium']
Smoking_Status_level_ord -> ['Current', 'Former', 'Never']
Alcohol_Consumption_Level_ord -> ['Occasionally', 'Never', 'Regularly']
Cholesterol_Level_ord -> ['Normal', 'High']
Depression_Level_ord -> ['Medium', 'Low', 'High']
Sleep_Quality_Level_ord -> ['Poor', 'Good', 'Average']
Diet_level_ord -> ['Healthy', 'Average', 'Unhealthy']
Air_Pollution_Exposure_level_ord -> ['High', 'Medium', 'Low']
Social_Engagement_Level_ord -> ['High', 'Low', 'Medium']
Income_Level_ord -> ['High', 'Medium', 'Low']
Stress_Levels_ord -> ['High', 'Medium', 'Low']


Cot target


In [ ]:
print(f"{target_col} -> {df[target_col].unique().tolist()}")


do_get_Alzheimer -> ['Yes', 'No']


Fill missing value


In [ ]:
cat_cols = binary_cols + nominal_cols + ordinal_cols + [target_col]

preprocessor = ColumnTransformer(
    transformers=[
        ("num", SimpleImputer(strategy="mean"), numeric_cols),
        ("cat", SimpleImputer(strategy="most_frequent"), cat_cols),
    ]
)

df_filled = preprocessor.fit_transform(df)
df_filled = pd.DataFrame(df_filled, columns=numeric_cols + cat_cols)

# Luu preprocessor lai
folder = "../artifacts/data_correction"
os.makedirs(folder, exist_ok=True)
myfuncs.save_python_object(
    file_path=os.path.join(folder, "missing_value_handler.pkl"), obj=preprocessor
)

df_filled.head()


Age_num Education_Level_num BMI_num Cognitive_Test_Score_num  \
0    52.0                 6.0    21.4                     67.0   
1    62.0                 4.0    28.7                     73.0   
2    52.0                19.0    21.3                     91.0   
3    74.0                15.0    32.4                     70.0   
4    94.0                17.0    33.9                     76.0   

  do_get_Diabetes_bin do_get_Hypertension_bin do_Family_History_bin  \
0                  No                      No                    No   
1                  No                     Yes                   Yes   
2                  No                      No                    No   
3                  No                      No                    No   
4                  No                     Yes                   Yes   

  do_get_Genetic_Risk_Factor_bin   Country_nom Gender_nom  ...  \
0                            Yes        France     Female  ...   
1                             No         Spain     Female  ...   
2                             No     Australia       Male  ...   
3                             No        Norway       Male  ...   
4                             No  Saudi Arabia     Female  ...   

  Alcohol_Consumption_Level_ord Cholesterol_Level_ord Depression_Level_ord  \
0                  Occasionally                Normal               Medium   
1                  Occasionally                Normal                  Low   
2                  Occasionally                  High                  Low   
3                         Never                Normal               Medium   
4                     Regularly                  High               Medium   

  Sleep_Quality_Level_ord Diet_level_ord Air_Pollution_Exposure_level_ord  \
0                    Poor        Healthy                             High   
1                    Good        Average                             High   
2                    Good      Unhealthy                           Medium   
3                 Average        Average                              Low   
4                    Good      Unhealthy                           Medium   

  Social_Engagement_Level_ord Income_Level_ord Stress_Levels_ord  \
0                        High             High              High   
1                         Low           Medium            Medium   
2                        High           Medium               Low   
3                         Low             High              High   
4                      Medium             High              High   

  do_get_Alzheimer  
0              Yes  
1               No  
2               No  
3               No  
4              Yes  

[5 rows x 25 columns]

Chuyen doi ve dung kieu du lieu


In [ ]:
df_filled[numeric_cols] = df_filled[numeric_cols].astype("float")


df_filled[cat_cols] = df_filled[cat_cols].astype("category")


df_filled.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66854 entries, 0 to 66853
Data columns (total 25 columns):
 #   Column                            Non-Null Count  Dtype   
---  ------                            --------------  -----   
 0   Age_num                           66854 non-null  float64 
 1   Education_Level_num               66854 non-null  float64 
 2   BMI_num                           66854 non-null  float64 
 3   Cognitive_Test_Score_num          66854 non-null  float64 
 4   do_get_Diabetes_bin               66854 non-null  category
 5   do_get_Hypertension_bin           66854 non-null  category
 6   do_Family_History_bin             66854 non-null  category
 7   do_get_Genetic_Risk_Factor_bin    66854 non-null  category
 8   Country_nom                       66854 non-null  category
 9   Gender_nom                        66854 non-null  category
 10  Employment_Status_nom             66854 non-null  category
 11  Marital_Status_nom                66854 non-null  cate

In [ ]:
df_filled.columns


Index(['Age_num', 'Education_Level_num', 'BMI_num', 'Cognitive_Test_Score_num',
       'do_get_Diabetes_bin', 'do_get_Hypertension_bin',
       'do_Family_History_bin', 'do_get_Genetic_Risk_Factor_bin',
       'Country_nom', 'Gender_nom', 'Employment_Status_nom',
       'Marital_Status_nom', 'living_place_urban_or_rural_nom',
       'Physical_Activity_Level_ord', 'Smoking_Status_level_ord',
       'Alcohol_Consumption_Level_ord', 'Cholesterol_Level_ord',
       'Depression_Level_ord', 'Sleep_Quality_Level_ord', 'Diet_level_ord',
       'Air_Pollution_Exposure_level_ord', 'Social_Engagement_Level_ord',
       'Income_Level_ord', 'Stress_Levels_ord', 'do_get_Alzheimer'],
      dtype='object')

In [ ]:
ord_values_dict = {
    "do_get_Diabetes_bin": ["No", "Yes"],
    "do_get_Hypertension_bin": ["No", "Yes"],
    "do_Family_History_bin": ["No", "Yes"],
    "do_get_Genetic_Risk_Factor_bin": ["No", "Yes"],
    "Physical_Activity_Level_ord": ["Low", "Medium", "High"],
    "Smoking_Status_level_ord": ["Never", "Former", "Current"],
    "Alcohol_Consumption_Level_ord": ["Never", "Occasionally", "Regularly"],
    "Cholesterol_Level_ord": ["Normal", "High"],
    "Depression_Level_ord": ["Low", "Medium", "High"],
    "Sleep_Quality_Level_ord": ["Poor", "Average", "Good"],
    "Diet_level_ord": ["Unhealthy", "Average", "Healthy"],
    "Air_Pollution_Exposure_level_ord": ["Low", "Medium", "High"],
    "Social_Engagement_Level_ord": ["Low", "Medium", "High"],
    "Income_Level_ord": ["Low", "Medium", "High"],
    "Stress_Levels_ord": ["Low", "Medium", "High"],
    "do_get_Alzheimer": ["No", "Yes"],
}

for col, value in ord_values_dict.items():
    df_filled[col] = df_filled[col].cat.reorder_categories(value, ordered=True)


In [ ]:
df_filled["do_get_Alzheimer"].cat.categories


Check duplicates


In [ ]:
any(df_filled.duplicated())


False

Khong co duplicates


In [ ]:
df_filled.shape


(66854, 25)

Chia thanh tap train va tap val


In [ ]:
df_train, df_val = train_test_split(
    df_filled, test_size=0.2, stratify=df[target_col], random_state=42
)

myfuncs.save_python_object(os.path.join(folder, "train_data.pkl"), df_train)
myfuncs.save_python_object(os.path.join(folder, "val_data.pkl"), df_val)
